File: chapter13/stringmatching_singlerow/C++/stringmatching.cpp
#include <string>
#include <iostream>

#define DIM 10000

using namespace std;

int D[DIM];

int stringMatching(string P, string T) {
  for (int j=0; j<=T.size(); j++) D[j] = 0;

  for (int i=1; i<=P.size(); i++) {
    int dx = i-1;
    for (int j=1; j<=T.size(); j++) {
      int t = min(min(dx + (P[i-1] == T[j-1] ? 0 : 1), D[j] + 1), D[j-1] + 1);

      dx = D[j];
      D[j] = t;
    }
    D[0] = i;
  }

  int minimo = D[0];
  for (int i=1; i<=T.size(); i++) {
    cout << D[i] << endl;
    minimo = min(minimo, D[i]);
  }

  return minimo;
}


int main() {
  string pattern("pippo");
  string text("pappa");

  cout << "Result: " << stringMatching(pattern, text) << endl;
}

File: chapter13/stringmatching_singlerow/Java/StringMatching.java
/*
 * Chapter 13
 * Paragraph 4
 */

public class StringMatching {
    final int DIM = 10000;
    private int D[] = new int[DIM];

    int match(String P, String T) {
        for (int j=0; j<=T.length(); j++) D[j] = 0;

        for (int i=1; i<=P.length(); i++) {
            int dx = i-1;
            for (int j=1; j<=T.length(); j++) {
                int t = Math.min(
                    Math.min(
                        dx + (P.charAt(i-1) == T.charAt(j-1) ? 0 : 1),
                        D[j] + 1
                    ),
                    D[j-1] + 1
                );

                dx = D[j];
                D[j] = t;
            }
            D[0] = i;
        }

        int minimo = D[0];
        for (int i=1; i<=T.length(); i++) minimo = Math.min(minimo, D[i]);

        return minimo;
    }


    public static void main(String[] args) {
        System.out.println("Result: " + new StringMatching().match("pippo", "voglio la pappa"));
    }
}
